x/executionlayer/handler.go

```diff
@@ -64,45 +67,87 @@ func handlerMsgCreateValidator(ctx sdk.Context, k ExecutionLayerKeeper, msg type
 
 	k.SetValidator(ctx, eeAddress, validator)
 
-	return getResult(true, msg)
+	return getResult(true, "")
 }
 
 func handlerMsgBond(ctx sdk.Context, k ExecutionLayerKeeper, msg types.MsgBond) sdk.Result {
-	err := k.Execute(ctx, msg.BonderAddress, msg.TokenContractAddress, msg.SessionCode, msg.SessionArgs, msg.PaymentCode, msg.PaymentArgs, msg.GasPrice)
-	if err != nil {
-		return getResult(false, msg)
-	}
-	return getResult(true, msg)
+	result, log := execute(ctx, k, msg.MsgExecute)
+	return getResult(result, log)
 }
 
 func handlerMsgUnBond(ctx sdk.Context, k ExecutionLayerKeeper, msg types.MsgUnBond) sdk.Result {
-	err := k.Execute(ctx, msg.UnbonderAddress, msg.TokenContractAddress, msg.SessionCode, msg.SessionArgs, msg.PaymentCode, msg.PaymentArgs, msg.GasPrice)
+	result, log := execute(ctx, k, msg.MsgExecute)
+
+	return getResult(result, log)
+}
+
+func execute(ctx sdk.Context, k ExecutionLayerKeeper, msg types.MsgExecute) (bool, string) {
+
+	// Parameter preparation
+	stateHash := ctx.CandidateBlock().State
+	protocolVersion := k.MustGetProtocolVersion(ctx)
+	log := ""
+
+	// Execute
+	deploys := []*ipc.DeployItem{}
+	deploy := util.MakeDeploy(msg.ExecAddress.ToEEAddress(), msg.SessionCode, msg.SessionArgs, msg.PaymentCode, msg.PaymentArgs, msg.GasPrice, ctx.BlockTime().Unix(), ctx.ChainID())
+	deploys = append(deploys, deploy)
+	reqExecute := &ipc.ExecuteRequest{
+		ParentStateHash: stateHash,
+		BlockTime:       uint64(ctx.BlockTime().Unix()),
+		Deploys:         deploys,
+		ProtocolVersion: &protocolVersion,
+	}
+	resExecute, err := k.client.Execute(ctx.Context(), reqExecute)
 	if err != nil {
-		return getResult(false, msg)
+		return false, err.Error()
+	}
+
+	effects := []*transforms.TransformEntry{}
+	switch resExecute.GetResult().(type) {
+	case *ipc.ExecuteResponse_Success:
+		for _, res := range resExecute.GetSuccess().GetDeployResults() {
+			switch res.GetExecutionResult().GetError().GetValue().(type) {
+			case *ipc.DeployError_GasError:
+				err = types.ErrGRpcExecuteDeployGasError(types.DefaultCodespace)
+			case *ipc.DeployError_ExecError:
+				err = types.ErrGRpcExecuteDeployExecError(types.DefaultCodespace, res.GetExecutionResult().GetError().GetExecError().GetMessage())
+			}
+
+			effects = append(effects, res.GetExecutionResult().GetEffects().GetTransformMap()...)
```

Comment:
Is it okay to append effect in case of GasError or ExecError?

Reply:
Yes, to prevent dos attacks. For exec error, gas is consumed and data is not reflected.